crates/outbound-redis/src/lib.rs

```diff
@@ -24,42 +30,87 @@ impl HostComponent for OutboundRedis {
-    ) -> anyhow::Result<Self::State> {
-        Ok(Self)
+    fn build_state(&self, component: &spin_manifest::CoreComponent) -> anyhow::Result<Self::State> {
+        let mut conn_map = HashMap::new();
+        if let Some(address) = component.wasm.environment.get("REDIS_ADDRESS") {
+            let client = redis::Client::open(address.to_string())?;
+            let conn = client.get_connection()?;
+            conn_map.insert(address.to_owned(), Mutex::new(conn));
+        }
+        Ok(Self {
+            connection: Arc::new(RwLock::new(conn_map)),
+        })
     }
 }
 
 impl outbound_redis::OutboundRedis for OutboundRedis {
     fn publish(&mut self, address: &str, channel: &str, payload: &[u8]) -> Result<(), Error> {
-        let client = redis::Client::open(address).map_err(|_| Error::Error)?;
-        let mut pubsub_conn = client.get_connection().map_err(|_| Error::Error)?;
-        pubsub_conn
-            .publish(channel, payload)
+        let conn_map = self.build_conn_and_get_conn_map(address)?;
+        let mut conn = conn_map
+            .get(address)
+            .unwrap()
+            .lock()
             .map_err(|_| Error::Error)?;
+        conn.publish(channel, payload).map_err(|_| Error::Error)?;
         Ok(())
     }
 
     fn get(&mut self, address: &str, key: &str) -> Result<Vec<u8>, Error> {
-        let client = redis::Client::open(address).map_err(|_| Error::Error)?;
-        let mut conn = client.get_connection().map_err(|_| Error::Error)?;
+        let conn_map = self.build_conn_and_get_conn_map(address)?;
+        let mut conn = conn_map
+            .get(address)
+            .unwrap()
+            .lock()
+            .map_err(|_| Error::Error)?;
         let value = conn.get(key).map_err(|_| Error::Error)?;
         Ok(value)
     }
 
     fn set(&mut self, address: &str, key: &str, value: &[u8]) -> Result<(), Error> {
-        let client = redis::Client::open(address).map_err(|_| Error::Error)?;
-        let mut conn = client.get_connection().map_err(|_| Error::Error)?;
+        let conn_map = self.build_conn_and_get_conn_map(address)?;
+        let mut conn = conn_map
+            .get(address)
+            .unwrap()
+            .lock()
+            .map_err(|_| Error::Error)?;
         conn.set(key, value).map_err(|_| Error::Error)?;
         Ok(())
     }
 
     fn incr(&mut self, address: &str, key: &str) -> Result<i64, Error> {
-        let client = redis::Client::open(address).map_err(|_| Error::Error)?;
-        let mut conn = client.get_connection().map_err(|_| Error::Error)?;
+        let conn_map = self.build_conn_and_get_conn_map(address)?;
+        let mut conn = conn_map
+            .get(address)
+            .unwrap()
+            .lock()
+            .map_err(|_| Error::Error)?;
         let value = conn.incr(key, 1).map_err(|_| Error::Error)?;
         Ok(value)
     }
 }
+
+impl OutboundRedis {
+    fn build_conn_and_get_conn_map<'ret, 'me: 'ret, 'c>(
```

Comment:
Could we have a slightly simpler name for this function?

Reply:
I changed it to `get_reused_conn_map`. I tried to let users know we set up a connection if we can't find it, but unsure how to describe it in a short term.